### PR TITLE
Krylov subspace expansion (with TDVP in min)

### PR DESCRIPTION
**Pull request by sajantanand**
Johannes, here is a rough version of what we discussed this morning. There is a function in `mps.py` that "extends" an MPS (in place) by additional provided MPS (or by random orthogonal vectors if none are provided). This required an LQ decomposition (from transposing the QR you already implemented) and truncated version of an eigenvalue decomposition.

I added this to TDVP in the `prepare_evolve` function in the base TDVP class. So before each L->R->L TDVP sweep, one can optionally expand the current MPS using either "Krylov" vectors generated by a specified MPO (or the MPO used for TDVP evolution) or random vectors. Then one can regular TDVP after this. One added annoyance is that since the bond dimensions of the wavefunction being evolved have changed, we need to clear out the LPs and RPs from the engine environment and recalculate them from scratch.

Questions:
1. Is it possible to update the environments such that we don't have to recalculate them from scratch? I don't see how we can do better than just starting at the right and contracting the new RPs.
2. Currently each time we extend by Krylov vectors, we load a truncation parameters dictionary for use in the eigenvalue decomposition function. This generates lots of logging output, since each loading of the dictionary generates something like
`INFO:tenpy.tools.params:truncation: reading 'chi_max'=None
INFO:tenpy.tools.params:truncation: reading 'chi_min'=None
INFO:tenpy.tools.params:truncation: reading 'degeneracy_tol'=None
INFO:tenpy.tools.params:truncation: reading 'svd_min'=0.01
INFO:tenpy.tools.params:truncation: reading 'trunc_cut'=1e-14`

I've checked with some small systems with exact bond dimension that single-site TDVP with subspace expansion produces results consistent with two-site TDVP. I will do more stringent testing (including using W2 to generate the Krylov vectors) before this should be merged, but if you have any improvements, please let me know.

### CODE REVIEW

**Comment by jhauschild** [comment]
1. I don't think that's possible. Also, calculating those Krylov vectors is significantly more expensive than recalculating the environments anyways...
2. You can avoid that if you use `trunc_params  = Krylov_options.subconfig('trunc_params')`  instead of `Krylov_options.get('trunc_params', ...)`. See also comments in the review.

**Comment by jhauschild** [comment]
Adding this here to make it more visible and make discussions easier:
I don't like the proposed `eig_theta` function.
Typical use cases will indeed be plugging in `rho`, where you know it's hermitian, not `theta` - that's also how you use it. 

One subtlety here is that you actually need different truncation - the `truncate` function truncates based on singular values, but eigenvalues of rho = |theta><theta| are S^2 as eigenvalues, so having the same `svd_min` will truncate singular values of the original theta as big as `sqrt(svd_min)` - I bet you didn't think of that, and surely the users won't...

Further, when `hermitian=False`, the eigenvectors of theta are not orthogonal, i.e. `U U^dagger != eye`,  `U^dagger != U^-1`, so your comment `theta ~= tensordot(V.scale_axis(W*renormalization, 1), V.conj().T, axes=1)` is wrong for hermitian=False.
In other words: please add a test function that checks the claims you make here ;-)

If you want to keep this function, call it `eigh_rho(...)`, discard the `hermitian=False` case (unless you really need it?) and call the truncate function based on sqrt(W).

Alternatively, I think you could replace it all-together if you work on the the level of individual B matrices directly instead of the density matrices, and replace using the projector into the null space with a gram-schmidt orthogonalizaiton against the existing columns of B, and adding the density matrices with stacking columns. Then the final step should indeed just be an svd_theta. (I didn't fully work this out, but I'm pretty confident that it should be possible.)

**Review by jhauschild — CHANGES_REQUESTED** [review]
a bunch of comments. The big change requested is restructuring `eig_theta` to and `eigh_rho` with different truncation scheme, the rest is rather minor cosmetics/improvements

File: tenpy/networks/mps.py [hunk 1]
```diff
@@ -4403,6 +4404,92 @@ def add(self, other, alpha, beta, cutoff=1.e-15):
...
+        Parameters
+        ----------
+        others : list of :class:`MPS` or an empty list ([])
+            MPS of same length used to extend the basis of self
+            If None, we add random basis states.
+        trunc_par : dict
+            Parameters for truncation, see :cfg:config:`truncation`.
+        do_canonicalize : Boolean
+            Do we convert to canonical form at the end? This has the tendency to remove some trivial
+            singular values which negates the purpose of this function.
```
[the remark this answers is not in the document]
Comment (sajantanand) [reply tenpy/networks/mps.py:4427]: The reason why I left it is that if one wants to have a "sane" MPS in memory after this function (i.e. one where the SVs on each bond have the proper length, where `psi.chi` actually returns the correct bond dimension, and where `psi.norm_test()` is all zeros), one should call canonical form. When using Krylov vectors to expand the basis, I am finding that canonicalizing the new MPS doesn't remove all of the new SVs that were added since the cutoff is 0.

[the remark this answers is not in the document]
Comment (sajantanand) [reply tenpy/networks/mps.py:4427]: What I can do is remove this option but clearly document to the user that the MPS is not canonical and ought to be used in a L->R->L sweep (i.e. DMRG or TDVP), where the sweeping will fix the MPS structure.

[the remark this answers is not in the document]
Comment (sajantanand) [reply tenpy/networks/mps.py:4427]: And your comment above about expanding the SVs on the bonds with zeros will ensure that `psi.chi` returns sensible values.

File: tenpy/algorithms/tdvp.py [hunk 2]
```diff
@@ -116,7 +156,7 @@ def evolve(self, N_steps, dt):
         N_steps : int
             The number of steps to evolve.
         """
-        consistency_check(dt, self.options, 'max_dt', 1.,
+        consistency_check(dt, self.options, 'max_dt', 2.,
```
[the remark this answers is not in the document]
Comment (sajantanand) [reply tenpy/algorithms/tdvp.py:159]: Jutho's original paper (https://arxiv.org/abs/1103.0936) states that TDVP has no Trotter error, but I believe this is the original formulation that does global updates rather than the more stable local update, sweeping method he introduced later (https://arxiv.org/abs/1408.5056). So the L->R->L sweeping method still has Trotter error even though one technically evolves by the full Hamiltonian (just projected into the one- or two-site variational space)?

[the remark this answers is not in the document]
Comment (sajantanand) [reply tenpy/algorithms/tdvp.py:159]: Is there a reason then that we don't implement a 4th order TDVP timestep that is directly analogous to a 4th order imaginary TEBD sweep?

[the remark this answers is not in the document]
Comment (sajantanand) [reply tenpy/algorithms/tdvp.py:159]: I was under the impression that if the bond dimension is large enough so that the projection of the Hamiltonian is accurate, then there is no discretization / Trotter error in TDVP, so I must be wrong.

File: tenpy/algorithms/truncation.py [hunk 3]
```diff
@@ -307,6 +307,66 @@ def svd_theta(theta, trunc_par, qtotal_LR=[None, None], inner_labels=['vR', 'vL'
...
+        The first label is inherited from `A`, the second label is ``'eig'``.
+    err : :class:`TruncationError`
+        The truncation error introduced.
+    """
+    if hermitian:
+        W, V = npc.eigh(theta, UPLO=UPLO, sort=sort)
+    else:
+        W, V = npc.eig(theta, sort=sort)
+
+    renormalization = np.linalg.norm(W)
```
[the remark this answers is not in the document]
Comment (sajantanand) [reply tenpy/algorithms/truncation.py:350]: Ah, yes. My bad.

File: tenpy/networks/mps.py [hunk 4]
```diff
@@ -4403,6 +4404,92 @@ def add(self, other, alpha, beta, cutoff=1.e-15):
...
+
+        current_Cs = []
+        for i in range(len(psis)):
+            current_Cs.append(psis[i].get_theta(L-1, n=1).replace_label('p0', 'p'))
+    
+        for j in reversed(range(1, L)):
+            _, exact_B = npc.lq(current_Cs[0].combine_legs(['p', 'vR']), mode='reduced', inner_labels=['vR', 'vL'])
+            eBhc_eB = npc.tensordot(exact_B.conj(), exact_B, axes=(['vL*'], ['vL']))
+            proj = npc.eye_like(eBhc_eB, labels=eBhc_eB.get_leg_labels()) - eBhc_eB
+            rho = npc.eye_like(proj, labels=eBhc_eB.get_leg_labels()) * 0
```
[the remark this answers is not in the document]
Comment (sajantanand) [reply tenpy/networks/mps.py:4455]: I didn't see this on the online documentation.  Is there a reason why it's not there? Maybe the doc string needs to be modified?

File: tenpy/networks/mps.py [hunk 5]
```diff
@@ -4403,6 +4404,92 @@ def add(self, other, alpha, beta, cutoff=1.e-15):
         psi.canonical_form_finite(renormalize=False, cutoff=cutoff)
         return psi
 
+    def extend(self, others=[], trunc_par={'svd_min': 1.e-8}, do_canonicalize=False):
```
[the remark this answers is not in the document]
Comment (sajantanand) [reply tenpy/networks/mps.py:4407]: I went with the name `subspace_expansion`. For what it's worth, the iTensor folk recently added the same function, and they call it `expand` (https://github.com/ITensor/ITensorTDVP.jl/blob/main/src/expand.jl).

File: tenpy/linalg/np_conserved.py [hunk 6]
```diff
@@ -4087,6 +4088,27 @@ def qr(a,
     return q, r
 
 
+def lq(a,
```
[the remark this answers is not in the document]
Comment (sajantanand) [reply tenpy/linalg/np_conserved.py:4091]: I don't think I understand. I need some decomposition that leave the unitary matrix on the right. LQ seemed the best choice since it's cheaper than an SVD and just involves QR + transposes.

[the remark this answers is not in the document]
Comment (sajantanand) [reply tenpy/linalg/np_conserved.py:4091]: I could use `polar` with `left=True`, but this works by doing an SVD, which I'm trying to avoid for efficiency reasons.

File: tenpy/algorithms/truncation.py [hunk 7]
```diff
@@ -307,6 +307,66 @@ def svd_theta(theta, trunc_par, qtotal_LR=[None, None], inner_labels=['vR', 'vL'
...
+        Usually, `theta` represents the wavefunction.
+    trunc_par : dict
+        truncation parameters as described in :func:`truncate`.
+    UPLO : {'L', 'U'}
+        Whether to take the lower ('L', default) or upper ('U') triangular part of `a`.
+        Only used for hermitian eigenvalue decomposition.
+    sort : {'m>', 'm<', '>', '<', ``None``}
+        How the eigenvalues should are sorted *within* each charge block.
+        Defaults to ``None``, which is same as '<'. See :func:`argsort` for details.
+    hermitian : Boolean
```
[the remark this answers is not in the document]
Comment (sajantanand) [reply tenpy/algorithms/truncation.py:332]: I've removed this argument now.

**Comment by sajantanand** [comment]
> 1. I don't think that's possible. Also, calculating those Krylov vectors is significantly more expensive than recalculating the environments anyways...

Technically everything scales as `O(chi^3)`, but you are right that the generation of Krylov vectors involves factors of the MPO bond dimension and thus is dominant.

**Comment by sajantanand** [comment]
Thanks for the detailed feedback! This was very helpful.

> One subtlety here is that you actually need different truncation - the `truncate` function truncates based on singular values, but eigenvalues of rho = |theta><theta| are S^2 as eigenvalues, so having the same `svd_min` will truncate singular values of the original theta as big as `sqrt(svd_min)` - I bet you didn't think of that, and surely the users won't...

You are very correct. I thought about this very briefly before, only long enough to leave some terse comments about this. I have switched to truncating based on `sqrt(eigs)` to be consistent. It's also worth noting that I'd expect the user to have a much larger `svd_min` here than when evolving the state itself, as we can tolerate more error when expanding the basis.

> If you want to keep this function, call it eigh_rho(...), discard the hermitian=False case (unless you really need it?) and call the truncate function based on sqrt(W).

I have gone with this route and added a test function.

File: tenpy/networks/mps.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sajantanand) [reply tenpy/networks/mps.py:4455]: I am using this now.

File: tenpy/algorithms/truncation.py [hunk 8]
```diff
@@ -307,6 +307,66 @@ def svd_theta(theta, trunc_par, qtotal_LR=[None, None], inner_labels=['vR', 'vL'
     return U, S, VH, err, renormalization
 
 
+def eig_theta(theta, trunc_par, UPLO='L', sort=None, hermitian=True):
```
[the remark this answers is not in the document]
Comment (sajantanand) [reply tenpy/algorithms/truncation.py:310]: I've followed your comments.

**Comment by sajantanand** [comment]
Hi @jhauschild do you have additional feedback? I didn't "resolve" all of your requested changes as I had questions about some of them. But I want to merge this into my DMT branch of my private fork so that I can use TDVP subspace expansion for operator evolution, which is my desired use case. I thought it would be best to merge this into main and then pull the changes into my DMT branch, but I can just merge these into my branch as long as that doesn't mess up the git version control.
